apps/aecore/src/aec_chain_state.erl

```diff
@@ -0,0 +1,458 @@
+top_block_hash(?match_state(top_block_hash := X)) ->
+    X.
+
+-spec top_block(state()) -> 'undefined' | #block{}.
+top_block(?match_state(top_block_hash := undefined)) -> undefined;
+top_block(?match_state(top_block_hash := X) = State) ->
+    export_block(blocks_db_get(X, State), State).
+
+-spec insert_block(#block{}, state()) -> {'ok', state()} | {'error', any()}.
+insert_block(Block, ?assert_state() = State) ->
+    try {ok, internal_insert(wrap_block(Block), State)}
+    catch throw:?internal_error(What) -> {error, What}
+    end.
+
+-spec insert_header(#header{}, state()) -> {'ok', state()} | {'error', any()}.
+insert_header(Header, ?assert_state() = State) ->
+    try {ok, internal_insert(wrap_header(Header), State)}
+    catch throw:?internal_error(What) -> {error, What}
+    end.
+
+-spec get_block(binary(), state()) -> {'ok', #block{}} | 'error'.
+get_block(Hash, ?assert_state() = State) ->
+    case blocks_db_find(Hash, State) of
+        {ok, Node} ->
+            case node_is_block(Node) of
+                true  -> {ok, export_block(Node, State)};
+                false -> error
+            end;
+        error -> error
+    end.
+
+-spec get_header(binary(), state()) -> {'ok', #header{}} | 'error'.
+get_header(Hash, ?assert_state() = State) ->
+    case blocks_db_find(Hash, State) of
+        {ok, Internal} -> {ok, export_header(Internal)};
+        error -> error
+    end.
+
+-spec difficulty_at_top_block(state()) -> {'ok', float()} | {'error', atom()}.
+difficulty_at_top_block(?assert_state() = State) ->
+    case get_top_block_hash(State) of
+        undefined -> {error, no_top};
+        Hash -> total_difficulty_at_hash(Hash, State)
+    end.
+
+-spec difficulty_at_top_header(state()) -> {'ok', float()} | {'error', atom()}.
+difficulty_at_top_header(?assert_state() = State) ->
+    case get_top_header_hash(State) of
+        undefined -> {error, no_top};
+        Hash -> total_difficulty_at_hash(Hash, State)
+    end.
+
+-spec difficulty_at_hash(binary(), state()) -> {'ok', float()} | {'error', atom()}.
+difficulty_at_hash(Hash, ?assert_state() = State) ->
+    total_difficulty_at_hash(Hash, State).
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+-spec internal_error(_) -> no_return().
+
+internal_error(What) ->
+    throw(?internal_error(What)).
+
+get_top_header_hash(#{top_header_hash := H}) -> H.
+set_top_header_hash(H, State) when is_binary(H) -> State#{top_header_hash => H}.
+
+get_top_block_hash(#{top_block_hash := H}) -> H.
+set_top_block_hash(H, State) when is_binary(H) -> State#{top_block_hash => H}.
+
+%%%-------------------------------------------------------------------
+%%% Internal ADT for differing between blocks and headers
+%%%-------------------------------------------------------------------
+
+-record(node, { type    :: 'block' | 'header'
```

Comment:
For persistence reasons, would it perhaps be preferable to use maps here?

Reply:
Maybe. It depends on how we will do persistence and how we will handle code change. If we use mnesia or dets a record is probably easiest. For RocksDB or something else we would need to do some serialization and deserialization anyway. Records might be slightly more efficient here than records. Versioning of the data will have to be handled anyway.

I will look at persistence next and perhaps come back and change this if needed but for now I suggest we go with records.